File: 2023/solutions/day01-2.cpp
#include <iostream>
#include <fstream>
#include <regex>
#include <vector>
#include <algorithm>
#include <iterator>

int main() {
	std::ifstream input{ "2023/inputs/day01-input.txt" };
	if (!input.is_open()) std::cout << "file did not open\n";
	std::regex r{ "one|two|three|four|five|six|seven|eight|nine|\\d"};
	std::vector<std::string> numbers{"one", "two", "three", "four", "five", "six", "seven", "eight", "nine"};
	std::string s;
	int sum{};
	while (std::getline(input, s)) {
		std::string num;
		std::smatch m{};
		std::regex_search(s, m, r);
		std::string first = m[0];
		std::string last;
		while (std::regex_search(s, m, r)) {
			last = m[0];
			s = s.substr(1);
		}
		int f, l;
		if (first.size() > 1) {
			f = std::distance(numbers.begin(), std::find(numbers.begin(), numbers.end(), first)) + 1;
		} else {
			f = first[0] - '0';
		}
		if (last.size() > 1) {
			l = std::distance(numbers.begin(), std::find(numbers.begin(), numbers.end(), last)) + 1;
		} else {
			l = last[0] - '0';
		}
		sum += f*10 + l;
	}
	std::cout << sum << '\n';
}

File: haskell/2023/solutions/day01-2.hs
import Data.Char
import Data.List as L

isSuffix :: String -> String -> Bool
isSuffix s x = drop (length x - length s) x == s 

digits = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "0", "1", "2", "3", "4", "5", "6", "7", "8", "9"]
firstDigit :: String -> [String] -> Char
firstDigit [] _ = '0'
firstDigit all@(_:xs) ys
   | digit == Nothing = firstDigit xs ys
   | otherwise        = getDigit (maybe "0" id digit)
   where digit = find (\d -> isPrefixOf d all) ys
         getDigit x = if length x == 1 then x !! 0 else intToDigit (maybe 0 id (findIndex (==x) ys) + 1)

revDigits = ["eno", "owt", "eerht", "ruof", "evif", "xis", "neves", "thgie", "enin", "0", "1", "2", "3", "4", "5", "6", "7", "8", "9"]
lastDigit :: String -> [String] -> Char
lastDigit x = firstDigit (reverse x)

calibration :: String -> Int
calibration x = read [firstDigit x digits, lastDigit x revDigits]

solution :: String -> String
solution x = show $ sum $ L.map calibration (L.lines x)

main = do
       file <- readFile "../inputs/2023/day01-input.txt"
       putStrLn $ solution file
       